README
About
-----
prepaid-manager-applet aims to ease the handling of mobile internet connections
using GSM mobile prepaid cards on the GNOME Desktop. Such a SIM card can either
be in a mobile phone used as a modem, a usb 3g module (usb surf stick) or used
by the built in 3G chipset in your laptop/netbook.

* It allows you to check the current balance and to top up the credit.
* It uses ModemManager to talk to your GSM chipset.

Requirements
------------
* ModemManager with ussd support (>= 0.5)
* mobile-broadband-provider-info with top-up support (>= 20110319)
* GTK+ >= 3.0.0, Debian package: libgtk+-3-0
* PyGObject >= 2.28.3, Debian package: python-gobject
* DBus Python Bindings >= 0.83, Debian package: python-dbus

Build
-----
To build ppm use:

  ./configure
  make

Running
-------
If you don't want to install ppm on you system you can run it from the source
tree:

   src/prepaid-manager-applet.py

Project Page
------------
https://honk.sigxcpu.org/piki/projects/ppm

Contact
-------
Send comments, patches and suggestions to

Guido Günther <agx@sigxcpu.org>


NEWS
===============
Version 0.0.3.1
===============
* Provider selections fixes
* Add 'change' button to allow to reconfigure current provider
* Reschedule modem setup on modem errors
* Switch to xz tarballs

==============
Version 0.0.3
==============
* Better handle missing and disabled modems
* Better feedback when entering top up codes if
  m-b-p-i has length information for the ussd code.

===============
Version 0.0.2.1
===============
* Update to PyGObject 3.0

=============
Version 0.0.2
=============
* Use GDBus instead of dbus-glib
* Add icon
* Use main window with menu bar instead of dialog
* Don't fail when provider disappears from m-b-p-i


src/ppm/provider.py
# vim: set fileencoding=utf-8 :
#
# (C) 2010 Guido Guenther <agx@sigxcpu.org>
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

import logging

class ProviderError(Exception):
    def __init__(self, msg):
        self.msg = msg


class Provider(object):
    """
    Keeps the information on howto interact with a certain provider, that
    his howto top up the ballance and how to query the current ballance

    It doesn't keep any current balance information or similar since this is
    associated with an account (a user can have several SIM cards form the same
    provider)
    """

    def __init__(self, country, name):
        self.country = country
        self.name = name
        self.fetch_balance_cmds = {}
        self.top_up_cmds = {}
        logging.debug("New provider: %s, %s", country, name)

    def add_fetch_balance_cmd(self, cmd):
        self.fetch_balance_cmds.update(cmd)
        logging.debug("Adding balance check %s" % cmd)

    def add_top_up_cmd(self, cmd):
        self.top_up_cmds.update(cmd)
        logging.debug("Adding top up %s" % cmd)

    def has_fetch_balance_cmd(self):
        # Only USSD for now
        if self.fetch_balance_cmds.has_key('ussd'):
            return True
        else:
            return False

    def has_top_up_cmd(self):
        # Only USSD for now
        if self.top_up_cmds.has_key('ussd'):
            return True
        else:
            return False

    def get_top_up_code_length(self):
        """The length of the topup code"""
        if self.has_top_up_cmd():
            return self.top_up_cmds['ussd'][2]
        else:
            return 0

    @property
    def top_up_code_length(self):
        return self.get_top_up_code_length()

    def fetch_balance(self, mm, reply_func=None, error_func=None):
        if self.has_fetch_balance_cmd():
            mm.ussd_initiate (self.fetch_balance_cmds['ussd'],
                              reply_func=reply_func,
                              error_func=error_func)
            return True
        else:
            return False

    def top_up(self, mm, code, reply_func=None, error_func=None):
        if self.has_top_up_cmd():
            cmd = self.top_up_cmds['ussd'][0].replace(
                                                  self.top_up_cmds['ussd'][1],
                                                  code)
            logging.debug("Top up cmd: %s", cmd)
            mm.ussd_initiate (cmd, reply_func=reply_func, error_func=error_func)
            return True
        else:
            return False


src/ppm/modemproxy.py
# vim: set fileencoding=utf-8 :
#
# (C) 2010 Guido Guenther <agx@sigxcpu.org>
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

from gi.repository import GObject
from gi.repository import GLib
from gi.repository import Gio

import re

class ModemError(Exception):
    def __init__(self, msg):
        self.msg = msg

    def is_forbidden(self):
        return [False, True][self.msg.find("Operation not allowed") != -1]

    def is_disabled(self):
         return [False, True][self.msg.find("not enabled") != -1]


class ModemManagerProxy(GObject.GObject):
    """Interface to ModemManager DBus API
    @ivar request: current pending request to ModemManager
    @type request: string
    @ivar modem: dbus path of modem we're currently acting on
    @type modem: string
    """

    MM_DBUS_SERVICE='org.freedesktop.ModemManager'
    MM_DBUS_INTERFACE_MODEM_MANAGER='org.freedesktop.ModemManager'
    MM_DBUS_OBJECT_MODEM_MANAGER='/org/freedesktop/ModemManager'
    MM_DBUS_INTERFACE_MODEM='org.freedesktop.ModemManager.Modem'
    MM_DBUS_INTERFACE_MODEM_GSM_CARD='org.freedesktop.ModemManager.Modem.Gsm.Card'
    MM_DBUS_INTERFACE_MODEM_GSM_USSD='org.freedesktop.ModemManager.Modem.Gsm.Ussd'
    MM_DBUS_TIMEOUT = 5000
    MM_DBUS_FLAGS = (Gio.DBusProxyFlags.DO_NOT_LOAD_PROPERTIES |
                     Gio.DBusProxyFlags.DO_NOT_CONNECT_SIGNALS)
    # IMSIs are 14 or 15 digits
    IMSI_RE = r'\d{14,15}'

    __gsignals__ = {
        # Emitted when a request to MM starts
        'request-started':  (GObject.SignalFlags.RUN_FIRST, None,
                             [object]),
        # Emitted when a request has finished
        'request-finished': (GObject.SignalFlags.RUN_FIRST, None,
                             [object]),
        }

    def __init__(self):
        GObject.GObject.__init__(self)
        self.bus = Gio.bus_get_sync(Gio.BusType.SYSTEM, None)
        self.request = None
        self.reply_func = None
        self.error_func = None
        self.modem = None
        self.obj = None

    def set_modem(self, modem):
        self.modem = modem

    def mm_request(func):
        def wrapped_f( self, *args, **kw):
            self.request = "%s" % func.func_name
            if kw.has_key('reply_func'):
                self.reply_func = kw['reply_func']
            if kw.has_key('error_func'):
                self.error_func = kw['error_func']
            self.emit('request-started', self)
            func(self, *args, **kw)
        wrapped_f.__name__= func.__name__
        wrapped_f.__doc__= func.__doc__
        return wrapped_f

    def mm_request_done(func):
        def wrapped_f( self, *args, **kw):
            self.emit('request-finished', self)
            ret = func(self, *args, **kw)
            self.reply_func = None
            self.error_func = None
            self.request = None
            return ret
        wrapped_f.__name__= func.__name__
        wrapped_f.__doc__= func.__doc__
        return wrapped_f

    def request_pending(self):
        if self.request:
            return True
        else:
            return False

    @mm_request_done
    def handle_dbus_reply(self, obj, result, user_data):
        try:
            res = obj.call_finish(result)
            if self.reply_func:
                self.reply_func(res, user_data)
        except Exception as err:
            if self.error_func:
                me = ModemError("%s failed: %s" % (self.request, err))
                self.error_func(me)

    def get_modems(self):
        modems = []
        mm = Gio.DBusProxy.new_sync(self.bus,
                                    self.MM_DBUS_FLAGS,
                                    None,
                                    self.MM_DBUS_SERVICE,
                                    self.MM_DBUS_OBJECT_MODEM_MANAGER,
                                    self.MM_DBUS_INTERFACE_MODEM_MANAGER,
                                    None)
        ret = mm.EnumerateDevices()
        for modem in ret:
            modems.append(modem)
        return modems

    def get_imsi(self):
        card = Gio.DBusProxy.new_sync(self.bus,
                                      self.MM_DBUS_FLAGS,
                                      None,
                                      self.MM_DBUS_SERVICE,
                                      self.modem,
                                      self.MM_DBUS_INTERFACE_MODEM_GSM_CARD,
                                      None)
        try:
            imsi = card.GetImsi()
        except Exception as msg:
            raise ModemError("Getting IMSI failed: %s" % msg)
        if not re.match(self.IMSI_RE, imsi):
            raise ModemError("%s is not a valid imsi" % imsi)
        return imsi

    def get_network_id(self):
        imsi = self.get_imsi()
        mcc = imsi[0:3]
        mnc = imsi[3:5]
        return (mcc, mnc)

    @mm_request
    def ussd_initiate(self, command, reply_func=None, error_func=None):
        ussd = Gio.DBusProxy.new_sync(self.bus,
                                      self.MM_DBUS_FLAGS,
                                      None,
                                      self.MM_DBUS_SERVICE,
                                      self.modem,
                                      self.MM_DBUS_INTERFACE_MODEM_GSM_USSD,
                                      None)
        ussd.call("Initiate", GLib.Variant('(s)', (command,)),
                  Gio.DBusCallFlags.NO_AUTO_START, self.MM_DBUS_TIMEOUT, None,
                  self.handle_dbus_reply, None)

    @mm_request
    def _modem__enable(self, enable, reply_func=None, error_func=None):
        ussd = Gio.DBusProxy.new_sync(self.bus,
                                      self.MM_DBUS_FLAGS,
                                      None,
                                      self.MM_DBUS_SERVICE,
                                      self.modem,
                                      self.MM_DBUS_INTERFACE_MODEM,
                                      None)
        ussd.call("Enable", GLib.Variant('(b)', (enable,)),
                  Gio.DBusCallFlags.NO_AUTO_START, self.MM_DBUS_TIMEOUT, None,
                  self.handle_dbus_reply, None)

    def modem_enable(self, reply_func=None, error_func=None):
        self._modem__enable(True,
                            reply_func=reply_func,
                            error_func=error_func)

    def modem_disable(self, reply_func=None, error_func=None):
        self._modem_enable(False,
                           reply_func=reply_func,
                           error_func=error_func)

GObject.type_register(ModemManagerProxy)


src/ppm/providerdb.py
# vim: set fileencoding=utf-8 :
#
# (C) 2010 Guido Guenther <agx@sigxcpu.org>
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

import os
import logging
from lxml import etree

from provider import Provider

class ProviderDB(object):
    """Proxy to mobile broadband provider database"""

    provider_info = os.getenv('PPM_PROVIDER_DB',
                              '/usr/share/mobile-broadband-provider-info/'
                              'serviceproviders.xml')
    country_codes = '/usr/share/zoneinfo/iso3166.tab'

    def __init__(self):
        self.__tree = None
        self.__countries = {}

    @property
    def tree(self):
        if self.__tree:
            return self.__tree
        else:
            self.__tree = etree.parse(self.provider_info)
            return self.__tree

    @property
    def countries(self):
        if self.__countries:
            return self.__countries
        else:
            self._tree = self._load_countries()
            return self.__countries

    def _load_countries(self):
        try:
            for line in file(self.country_codes, 'r'):
                if line[0] != '#':
                    (code, country) = line.split('\t',2)
                    self.__countries[code.lower()] = country.strip()
        except IOError as msg:
            logging.warning("Loading country code database failed: %s" % msg)

    def _fill_provider_info(self, provider_elem):
        """Fill a provider object with data from the XML"""
        name = provider_elem.xpath("./name")[0].text
        country = provider_elem.getparent().attrib['code']
        gsm_elem = provider_elem.xpath("./gsm")
        provider = Provider(country = country,
                            name = name)
        if gsm_elem:
            self._fill_balance_check_cmd(gsm_elem[0], provider)
            self._fill_top_up_cmd(gsm_elem[0], provider)
        return provider

    def _fill_balance_check_cmd(self, xmlelemnt, provider):
        """Fetch balance check method from XML and add it to the provider"""
        for child in xmlelemnt.iter(tag='balance-check'):
            check_types = child.getchildren()
            for t in check_types:
                if t.tag == 'ussd':
                    sequence = t.text
                    provider.add_fetch_balance_cmd({'ussd': sequence})
                if t.tag == 'sms':
                    number = t.text
                    text = t.attrib['text']
                    provider.add_fetch_balance_cmd({'sms': (number, text)})

    def _fill_top_up_cmd(self, xmlelement, provider):
        for child in xmlelement.iter(tag='balance-top-up'):
            check_types = child.getchildren()
            for t in check_types:
                if t.tag == 'ussd':
                    sequence = t.text
                    replacement = t.attrib['replacement']
                    try:
                        length = int(t.attrib['length'])
                    except KeyError:
                        length = 0
                    provider.add_top_up_cmd({'ussd': [sequence,
                                                      replacement,
                                                      length]})
                if t.tag == 'sms':
                    number = t.text
                    text = t.attrib['text']
                    provider.add_top_up_cmd({'sms': (number, text)})

    def get_providers(self, mcc, mnc):
        """
        Get possible providers for the current mcc and mnc from the database
        """
        path = "//network-id[@mcc='%s' and @mnc='%s']" % (mcc, mnc)
        searcher = etree.ETXPath(path)
        providers = []

        try:
            for r in searcher(self.tree):
                provider_elem = r.getparent().getparent()
                providers.append(self._fill_provider_info(provider_elem))
        except etree.XMLSyntaxError:
            return None
        return providers

    def get_provider(self, country_code, name):
        path = "//country[@code='%s']/provider[name='%s']" % (country_code, name)
        searcher = etree.ETXPath(path)

        for r in searcher(self.tree):
            return self._fill_provider_info(r)
        return None

    def get_country_codes(self):
        path = "/serviceproviders/country"
        searcher = etree.ETXPath(path)

        for r in searcher(self.tree):
            yield r.attrib['code']

    def get_country_by_code(self, code):
        """Given a country code return it's name"""
        try:
            return self.countries[code]
        except KeyError:
            return None

    def get_countries(self):
        for code in self.get_country_codes():
            try:
                yield (self.countries[code], code)
            except KeyError:
                yield (None, code)

    def get_providers_by_code(self, country_code):
        path = ("/serviceproviders/country[@code='%s']/provider/name" %
                country_code)
        searcher = etree.ETXPath(path)

        for r in searcher(self.tree):
            yield r.text


src/ppm/accountdb.py
# vim: set fileencoding=utf-8 :
#
# (C) 2011 Guido Günther <agx@sigxcpu.org>
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

from gi.repository import Gio
from gi.repository import GObject
import logging

from provider import Provider

class Account(GObject.GObject):
    identifier = GObject.property(type=str,
                                  nick='account identifier',
                                  blurb='the identifier of this account, usually the imsi')
    name = GObject.property(type=str,
                            nick='provider name',
                            blurb='name of the provider')
    code = GObject.property(type=str,
                            nick='country name',
                            blurb='counry code the  provider is in')
    balance = GObject.property(type=str,
                               nick='current balance',
                               blurb='current balance of this account')
    timestamp = GObject.property(type=str,
                                 nick='update timestamp',
                                 blurb='last time the balance info got updated')

    def update_provider(self, provider):
        """Update the provider information"""
        if provider.country != self.props.code:
            self.props.code = provider.country
        if provider.name != self.props.name:
            self.props.name = provider.name

    def update_balance(self, balance, timestamp):
        """Update balance information"""
        if timestamp != self.props.timestamp:
            self.props.balance = balance
            self.props.timestamp = timestamp

GObject.type_register(Account)


class AccountDB(object):
    """
    In Gsettings we keep the detailed provider information that's associated
    with an account identified by the SIM card. A user can have several
    accounts at the same provider.
    """

    PPM_GSETTINGS_ID = 'org.gnome.PrepaidManager'
    PPM_GSETTINGS_ACCOUNT_ID = PPM_GSETTINGS_ID + '.account'

    def __init__(self):
        self.settings = Gio.Settings(self.PPM_GSETTINGS_ID)
        self.accounts_path_prefix = self.settings.get_property("path") + 'accounts/'

    def is_known_account(self, imsi):
        """Do we know about this account in GSettings?"""
        return self.imsi_to_identifier(imsi) in self.settings.get_strv('accounts')

    def _account_path(self, imsi):
        """
        This key identifies the provider setup for a specific SIM card in
        GSettings. Since we might not be allowed to get the imsi from the
        phone other keys must be possible later
        """
        return '%s%s/' % (self.accounts_path_prefix,
                         self.imsi_to_identifier(imsi))

    @classmethod
    def imsi_to_identifier(klass, imsi):
        """Turn an imsi into an identifier, this allows us to support non
        imsi based accounts later"""
        return 'imsi:%s' % imsi

    def _bind_account(self, imsi):
        """Bind a new account object to a gsettings path"""

        path = self._account_path(imsi)
        account = Account()
        account.props.identifier =  self.imsi_to_identifier(imsi)
        gsettings_account = Gio.Settings(self.PPM_GSETTINGS_ACCOUNT_ID, path)
        gsettings_account.bind('provider', account, 'name',
                                Gio.SettingsBindFlags.DEFAULT)
        gsettings_account.bind('country', account, 'code',
                                Gio.SettingsBindFlags.DEFAULT)
        gsettings_account.bind('balance', account, 'balance',
                                Gio.SettingsBindFlags.DEFAULT)
        gsettings_account.bind('timestamp', account, 'timestamp',
                                Gio.SettingsBindFlags.DEFAULT)
        return account

    def fetch(self, imsi):
        """Given an imsi check if we know about it"""

        if self.is_known_account(imsi):
            logging.debug("IMSI '%s' not known", imsi)
            return None
        else:
            logging.debug("Fetching account information from '%s'",
                          self._account_path(imsi))

        account = self._bind_account(imsi)
        if account.props.name and account.props.code:
            logging.debug("Provider '%s' in '%s'", account.props.name,
                          account.props.code)
            return account
        else:
            logging.debug("Account not yet known.")
            return None

    def add(self, imsi, provider):
        """Given an imsi and a provider persist a new account with  that
        information"""
        path = self._account_path(imsi)

        logging.debug("Persisting '%s' in '%s' at '%s'",
                      provider.name,
                      provider.country,
                      path)
        account = self._bind_account(imsi)
        account.props.name = provider.name
        account.props.code = provider.country


